WordPress/gutenberg
# Enhance attributes to specify types and defaults

Closes #865
Closes #973
Supersedes #892, #988
Related: #391, #714

This pull request seeks to refactor the block type `attributes` property to specify not only attributes parsed from content, but also those also encoded in the block comment delimiter. Further, it optionally allows specifying type and default value of an attribute, replacing previous efforts including `defaultAttributes` and render-time destructuring.

The existing behavior of `attributes` matchers will continue to work as-is, but comment attributes are no longer inferred and must instead be explicitly defined. The value of each key in `attributes` can be a matcher, a constructor (type), or an object specifying any of `type`, `matcher`, and `defaultValue`.

See the following comment for the "kitchen sink" example: https://github.com/WordPress/gutenberg/pull/1905#issuecomment-320159018

<p><details>
<summary><strike>See original kitchen sink example</strike></summary>
<pre><code>it( 'should merge attributes with the parsed and default attributes', () => {
	const blockType = {
		attributes: {
			content: text( 'div' ),
			number: {
				type: Number,
				matcher: attr( 'div', 'data-number' ),
			},
			align: String,
			topic: {
				type: String,
				defaultValue: 'none',
			},
			ignoredDomMatcher: {
				matcher: ( node ) => node.innerHTML,
			},
		},
	};

	const rawContent = '<div data-number="10">Ribs</div>';
	const attrs = { align: 'left', invalid: true };

	expect( getBlockAttributes( blockType, rawContent, attrs ) ).toEqual( {
		content: 'Ribs',
		number: 10,
		align: 'left',
		topic: 'none',
	} );
} );</code></pre>
</details></p>

~Block authors can potentially implement their own type coercion by specifying a constructor with `valueOf` implementation. There is some inspiration here from [Vue.js prop validation](https://vuejs.org/v2/guide/components.html#Prop-Validation) and [`prop-types`](https://www.npmjs.com/package/prop-types).~ (Removed)

__Testing instructions:__

Ensure that tests pass:

```
npm test
```

Verify that there is no regressions in block usage, particularly:

- Block default values
- Block transforms
- Block serialization of comment attributes
